File: ELEC881_Copilot/src/Original Prompt/Java script (JS)/q12-largest-divisible-subset.js
/*
Given a set of distinct positive integers nums, return the largest subset answer such that every pair (answer[i], answer[j]) of elements in this subset satisfies:

answer[i] % answer[j] == 0, or
answer[j] % answer[i] == 0
If there are multiple solutions, return any of them.

 

Example 1:

Input: nums = [1,2,3]
Output: [1,2]
Explanation: [1,3] is also accepted.
Example 2:

Input: nums = [1,2,4,8]
Output: [1,2,4,8]
 

Constraints:

1 <= nums.length <= 1000
1 <= nums[i] <= 2 * 109
All the integers in nums are unique.
*/
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var largestDivisibleSubset = function(nums) {
    nums.sort((a, b) => a - b);
    let dp = new Array(nums.length).fill(1);
    let max = 1;
    let maxIndex = 0;
    for (let i = 1; i < nums.length; i++) {
        for (let j = 0; j < i; j++) {
            if (nums[i] % nums[j] === 0) {
                dp[i] = Math.max(dp[i], dp[j] + 1);
                if (dp[i] > max) {
                    max = dp[i];
                    maxIndex = i;
                }
            }
        }
    }
    let res = [];
    let prev = nums[maxIndex];
    for (let i = maxIndex; i >= 0; i--) {
        if (prev % nums[i] === 0 && dp[i] === max) {
            res.unshift(nums[i]);
            prev = nums[i];
            max--;
        }
    }
    return res;
    
};

File: ELEC881_Tabnine/src/No Description Prompt/Python/q12-largest-divisible-subset.py
class Solution(object):
    def largestDivisibleSubset(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        nums.sort()
        dp = [1] * len(nums)
        prev = [-1] * len(nums)

        max_length = 0
        max_index = 0

        for i in range(1, len(nums)):
            for j in range(i):
                if nums[i] % nums[j] == 0 and dp[i] < dp[j] + 1:
                    dp[i] = dp[j] + 1
                    prev[i] = j

            if dp[i] > max_length:
                max_length = dp[i]
                max_index = i

        result = []
        while max_index != -1:
            result.append(nums[max_index])
            max_index = prev[max_index]

        return result[::-1]


        